Automatically wipe dependent fields on service "type" updates

thockin opened the pull request:
Service has had a problem since forever:

* User creates a service type=LoadBalancer
* We silently allocate them a NodePort
* User changes type to ClusterIP
* We fail the operation because they did not clear NodePort

They never asked for or used the NodePort!

This change introduces auto-wiping of dependent fields.  If you squint,
you can see Service as a big, messy discriminated union, with `type` as
the discriminator.  Ignoring fields for non-selected union-modes seems
right.

This introduces the potential for an apply loop.  Specifically, we will
accept YAML that we did not previously accept.  Apply *could* see the
field in local YAML and not in the server and repeatedly try to patch it
in.  But since that YAML is currently an error, it seems like a very low
risk.  Almost nobody actually specifies their own NodePort values.

To mitigat this somewhat, we only auto-wipe on updates.  The same YAML
would fail to create.  This is a little inconsistent.  We could
auto-wipe on create, too, at the risk of more potential impact.

To do this properly, we need to know the old and new values, which means
we can not do it in defaulting or conversion.  So we do it in strategy.


/kind bug
/kind api-change
/sig network
/sig api-machinery

**Which issue(s) this PR fixes**:

Fixes #33766

```release-note
Certain fields on  Service objects will be automatically cleared when changing the service's `type` to a mode that does not need those fields.  For example, changing from type=LoadBalancer to type=ClusterIP will clear the NodePort assignments, rather than forcing the user to clear them.
```

fedebongio commented:
/cc @lavalamp @liggitt

apelisse commented on pkg/registry/core/service/strategy.go:
```diff
@@ -205,6 +207,50 @@ func topologyKeysInUse(svc *api.Service) bool {
+//
+// Services are effectively a discriminated union, where type is the
+// discriminator. Some fields just don't make sense with some types.
+func normalizeLinkedFields(newSvc *api.Service, oldSvc *api.Service) {
```
Do we want the name to be more specific? That sounds too generic and I'm concerned people would want to insert all kind of changes here. Maybe `wipeNodePortsOnNewType` or something that is very targeted to what we're doing?

thockin replied:
Well, this is somewhat a trial baloon.  Dual-stack will have to add other fields here, too.  I could be convinced that `wipeIrrelevantFields` and `handleLinkedFields` are 2 different functions if you think that is important.  The main point is that something special is happening

apelisse commented on test/e2e/network/service.go:
```diff
@@ -1552,7 +1552,6 @@ var _ = SIGDescribe("Services", func() {
 		ginkgo.By("changing TCP service back to type=ClusterIP")
 		_, err = tcpJig.UpdateService(func(s *v1.Service) {
 			s.Spec.Type = v1.ServiceTypeClusterIP
-			s.Spec.Ports[0].NodePort = 0
```
Any test we could add to show the new behavior?

thockin replied:
Great point, will add verification

lavalamp replied:
Yeah still thinking about the change, but I think a unit test would be better?

thockin replied:
We have unit tests and this makes e2e also verify this behavior.

thockin commented:
Hopefully CI will pass this time.

The looming question: Am I going to regret this?  So far nobody has convinced me I will...

apelisse commented on pkg/registry/core/service/strategy.go:
```diff
@@ -205,6 +207,50 @@ func topologyKeysInUse(svc *api.Service) bool {
+}
+
+func needsClusterIP(svc *api.Service) bool {
+	if svc.Spec.Type == api.ServiceTypeExternalName {
+		return false
+	}
+	return true
```
I don't know how you feel about this in general:
```suggestion
	return svc.Spec.Type != api.ServiceTypeExternalName
```

apelisse commented on pkg/registry/core/service/strategy.go:
```diff
@@ -205,6 +207,50 @@ func topologyKeysInUse(svc *api.Service) bool {
+}
+
+func needsNodePort(svc *api.Service) bool {
+	if svc.Spec.Type == api.ServiceTypeNodePort || svc.Spec.Type == api.ServiceTypeLoadBalancer {
+		return true
+	}
+	return false
```
```suggestion
	return svc.Spec.Type == api.ServiceTypeNodePort || svc.Spec.Type == api.ServiceTypeLoadBalancer
```

apelisse commented:
Maybe rename PR title to "Automatically wipe dependent fields on Service type updates", otherwise it's very broad.

> Am I going to regret this? So far nobody has convinced me I will...

Almost certainly :-)

apelisse commented:
LGTM for me, I'll let someone else decide if that's a good idea.

lavalamp commented:
I have a bit of a queue, will look when I am able...

thockin commented:
This is a small change, but requires an opinion :)

lavalamp commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -4012,7 +4012,7 @@ type ServiceSpec struct {
 	// of the service will fail. This field can not be changed through updates.
```
How can you wipe it if it cannot be changed through updates?

liggitt replied:
it *can* be changed iff the old service or new service in the update request was of type ExternalName:

https://github.com/kubernetes/kubernetes/blob/2318a13228dfaadd2de4669c8ce3ffb126ed3958/pkg/apis/core/validation/validation.go#L4445-L4455

lavalamp commented:
Since you said it was small, I looked... 

I think this won't generate conflicts for SSA users with a wrong manifest, which would be nice?

If that's the worst thing that can be said for it, it's probably fine.

Maybe @liggitt can think of something bad that will happen.

thockin commented:
Verify fails for openapi, but running it locally doesn't update anything.  Sigh.  Every PR I write, I end up debugging something in hack

apelisse commented:
Yes, verify is correct, you've updated the comment which will update the description of the type in the openapi.

thockin commented:
Sure, but my local client is not seeing that - something is not being rebuilt when it needs to be.  The binary result of `make kube-apiserver` is still emitting "Ignored" rather than "Wiped".  I don't know enough about how those strings get into kube-apiserver and I don't have time to debug this build failure right now, so I just did `make clean` and it not finds that diff.

Yay.

thockin commented:
```
$ git grep clusterIP.*None.*Ignored
staging/src/k8s.io/cli-runtime/artifacts/openapi/swagger.json:          "description": "clusterIP is the IP address of the service and is usually assigned randomly by the master. If an address is specified manually and is not in use by others, it will be allocated to the service; otherwise, creation of the service will fail. This field can not be changed through updates. Valid values are \"None\", empty string (\"\"), or a valid IP address. \"None\" can be specified for headless services when proxying is not required. Only applies to types ClusterIP, NodePort, and LoadBalancer. Ignored if type is ExternalName. More info: https://kubernetes.io/docs/concepts/services-networking/service/#virtual-ips-and-service-proxies",
staging/src/k8s.io/kubectl/testdata/openapi/swagger.json:          "description": "clusterIP is the IP address of the service and is usually assigned randomly by the master. If an address is specified manually and is not in use by others, it will be allocated to the service; otherwise, creation of the service will fail. This field can not be changed through updates. Valid values are \"None\", empty string (\"\"), or a valid IP address. \"None\" can be specified for headless services when proxying is not required. Only applies to types ClusterIP, NodePort, and LoadBalancer. Ignored if type is ExternalName. More info: https://kubernetes.io/docs/concepts/services-networking/service/#virtual-ips-and-service-proxies",
```

How do THOSE get fixed?

apelisse commented:
I would expect these to be used either for test purposes (staging/src/k8s.io/kubectl/*testdata*/openapi) and don't need updating, or are updated manually occasionally. I think you're fine.

thockin commented:
/retest

thockin commented:
Re-pushed with renamed function and more comments

liggitt commented on pkg/registry/core/service/strategy.go:
```diff
@@ -205,6 +207,61 @@ func topologyKeysInUse(svc *api.Service) bool {
+	// "None" counts), we can wipe ClusterIP.  There are other fields that
+	// could also be wiped, but historically were ignored.  For compat, we
+	// leave it that way.
+	if needsClusterIP(oldSvc) && !needsClusterIP(newSvc) {
```
also make sure newSvc did not change the clusterIP (if the incoming request explicitly monkeyed with the field, they should fail rather than get wiped and succeed)

thockin replied:
ACK, in next push

liggitt commented on pkg/registry/core/service/strategy.go:
```diff
@@ -205,6 +207,61 @@ func topologyKeysInUse(svc *api.Service) bool {
+
+	// If the user is switching to a type that doesn't use NodePorts, we can
+	// wipe any allocated ports.
+	if needsNodePort(oldSvc) && !needsNodePort(newSvc) {
```
if possible limit clearing to fields we inherited unchanged from the old object

thockin replied:
ACK, in next push

liggitt commented on pkg/registry/core/service/strategy.go:
```diff
@@ -205,6 +207,61 @@ func topologyKeysInUse(svc *api.Service) bool {
+			newSvc.Spec.Ports[i].NodePort = 0
+		}
+	}
+
```
do we also allocate HealthCheckNodePort? how does that interact with load balancer policy. check what API validation existed around this case.

thockin replied:
That is cleared later, but I will add it here for completeness.

liggitt commented:
this seems like a good addition for the narrowly scoped cases where we auto-allocate these fields

can move this question to Kal's PR if you want, but I'm curious if/when this is expanded to defaulted dual-stack fields (ipFamilies, ipFamilyPolicy, etc), under what circumstances would you envision clearing/re-populating those? something like "if the old and new field for {ipFamily,ipFamilyPolicy,clusterIPs} match values, and that value is not valid for the new service type"?

liggitt commented on pkg/registry/core/service/strategy.go:
```diff
@@ -101,6 +101,7 @@ func (strategy svcStrategy) PrepareForCreate(ctx context.Context, obj runtime.Ob
 	}
 
 	dropServiceDisabledFields(service, nil)
+	dropTypeDependentFields(service, nil)
```
would you envision ipFamilies/ipFamilyPolicy/clusterIPs population happening *after* this call?

thockin replied:
Yes, allocation happens later

liggitt commented on pkg/registry/core/service/strategy.go:
```diff
@@ -119,6 +120,7 @@ func (strategy svcStrategy) PrepareForUpdate(ctx context.Context, obj, old runti
 	}
 
 	dropServiceDisabledFields(newService, oldService)
+	dropTypeDependentFields(newService, oldService)
```
same question about ordering w.r.t. ipFamilies/ipFamilyPolicy/clusterIPs population

thockin replied:
Yes, allocation happens later

thockin commented:
> something like "if the old and new field for {ipFamily,ipFamilyPolicy,clusterIPs} match values, and that value is not valid for the new service type"?

Exactly.  IF the user made a valid change of `type` we will clear fields that don't make sense for the new type (modulo compat, so not changing fields that don't today cause errors).

New push is up, now with a test.  PTAL

thockin commented:
Adding 2 more test cases, BRB

thockin commented:
Added ability to add/remove ports (valid in API) as long as that doesn't increase the set of NodePorts

thockin commented:
PTAL at your leisure

liggitt commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -4012,7 +4012,7 @@ type ServiceSpec struct {
 	// can be specified for headless services when proxying is not required.
-	// Only applies to types ClusterIP, NodePort, and LoadBalancer. Ignored if
+	// Only applies to types ClusterIP, NodePort, and LoadBalancer. Wiped if
 	// type is ExternalName.
```
should describe the exact circumstances under which it is cleared (e.g. "IP values are cleared on update from non-ExternalName service to ExternalName service if the value ClusterIP value remains unchanged")

thockin replied:
OK, a bunch of these comments can be updated.  Will do and re-push.

liggitt commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -4012,7 +4012,7 @@ type ServiceSpec struct {
 	// Valid values are "None", empty string (""), or a valid IP address. "None"
 	// can be specified for headless services when proxying is not required.
-	// Only applies to types ClusterIP, NodePort, and LoadBalancer. Ignored if
+	// Only applies to types ClusterIP, NodePort, and LoadBalancer. Wiped if
```
similar doc updates needed for healthz and nodeports

thockin commented on pkg/registry/core/service/strategy.go:
```diff
@@ -205,6 +207,127 @@ func topologyKeysInUse(svc *api.Service) bool {
+	// There are other fields that could also be wiped, but historically were
+	// ignored.  For compat, we leave it that way.
+	if needsClusterIP(oldSvc) && !needsClusterIP(newSvc) && sameClusterIP(oldSvc, newSvc) {
+		newSvc.Spec.ClusterIP = ""
```
yes - "None" is a value and is currently invalid when type=ExternalName, so this is consistent.  I mentioned it in comments above :)

thockin commented on pkg/registry/core/service/strategy.go:
```diff
@@ -205,6 +207,127 @@ func topologyKeysInUse(svc *api.Service) bool {
+func sameNodePorts(oldSvc, newSvc *api.Service) bool {
+	// Helper to make a unique key so we can tell if a port changed.
+	makeKey := func(p *api.ServicePort) string {
+		return fmt.Sprintf("%s/%d", p.Protocol, p.Port)
```
I thought about that and it seems correct.  It is effectively a new port definition.  So if you did this AND switch to ExternalName at the same time, it would fail (later in validation) for having a NodePort value.

We simply don't have an immutable key for ports, so I can't really tell the difference between "I changed an existing port number which already had a NodePort" and "I added a new port and specified a NodePort for it".  I think this is as close to correct as we can get.

There's a test-case for it "don't clear changed nodePorts" and "don't clear nodePorts when adding a port with NodePort"

liggitt replied:
seems reasonable, just wanted to make sure we considered it

liggitt commented:
lgtm modulo doc updates

thockin commented:
I fixed up a bunch of the API doc-comments.  PTAL

khenidak commented on pkg/registry/core/service/strategy.go:
```diff
@@ -205,6 +207,127 @@ func topologyKeysInUse(svc *api.Service) bool {
+	// Every old port must have the same NodePort in the new config or be
+	// cleared to 0 or must be removed.
+	for k, oldVal := range oldPorts {
+		if newVal, found := newPorts[k]; found {
```
What happens to partial remove, eg change type and remove one of n ports?

thockin replied:
If they remove a port it is fine.

If they change a port number or protocol, this treats it as if they removed the old and added the new.  If they did not clear NodePort, we will not clear it (they "added" a port with NodePort set) and will fail validation.

I thought it was important to allow adding a new port and removing an old port.  Changing an existing port is more complicated.  I could buy an argument that as long as the node port value didn't change, we could still wipe it (though that's not what is here).  Do you think that's better?  Now that I spell it out, I am coming around..

examples (assume {protocol, port, nodeport)


Good Add: New port does not have a NodePort
`{ type: NodePort, ports: [{TCP, 80, 30080}] }` -> `{ type: ClusterIP, ports: [{TCP, 80, 30080}, {TCP, 443}] }`

Bad Add: New port has a NodePort
`{ type: NodePort, ports: [{TCP, 80, 30080}] }` -> `{ type: ClusterIP, ports: [{TCP, 80, 30080}, {TCP, 443, 30443}] }`

Good Del:
`{ type: NodePort, ports: [{TCP, 80, 30080}, {TCP, 443, 30443}] }` -> `{ type: ClusterIP, ports: [{TCP, 80, 30080}] }`

Good Mod: New port does not have a NodePort
`{ type: NodePort, ports: [{TCP, 80, 30080} }` -> `{ type: ClusterIP, ports: [{TCP, 443}] }`

The open question is this one (set to demonstrate worst case):
`{ type: NodePort, ports: [{TCP, 80, 30080} }` -> `{ type: ClusterIP, ports: [{UDP, 53, 30080}] }`

Both the protocol and port number changed, but the NodePort stayed the same.  Can I wipe the node port safely?  I was thinking not, but now I am changing my thinking...

thockin replied:
New push is up as a 2nd commit.  Will mark this PR for squash

khenidak replied:
the original line of thought (or at least that is what i thought we are trying to do):
as captured by this line https://github.com/kubernetes/kubernetes/pull/95196/files#diff-2dd9a1ef6872092e519941c8a4870f4c2a71eda096b660510b02afe9b3d19fefR243

`user should be able to change a service type to a service that does not need NodePorts without having to wipe out NodePorts while patching the Service`

**a truth table would look like this** 
O = old
N = new
O-type-need-node-ports => N-type-need-node-ports + ports changed               `OK: if node port is immutable*`
O-type-need-node-ports => N-type-need-node-ports + ports same                     `OK: validate other fields`
O-type-need-node-ports => N-DO NOT-need-node-ports + ports same               `OK: clear intent`
O-type-need-node-ports => N-DO NOT-need-node-ports + ports cleared            `OK: clear intent`
O-type-need-node-ports => N-DO NOT-need-node-ports + ports changed `NOT OK: unclear intent **`
O-type-DO NOT- need-node-ports => N-DO NOT-need-node-ports + add ports `NOT OK: ports! (validation)`
O-type-DO NOT- need-node-ports => N-DO NOT-need-node-ports + no ports    `OK`
O-type-DO NOT- need-node-ports => N-need-node-ports + no ports                    `OK`
O-type-DO NOT- need-node-ports => N-need-node-ports + some ports `OK: if ports available (alloc)***`
O-type-DO NOT- need-node-ports => N-need-node-ports + all ports `OK: if ports available (alloc)***`

* i am not even sure if we allow changing them (i am assuming yes)
** any change add/change/partial delete did the user really want to change type or made a mistake?
*** assuming we allow specific port allocation if available (similar to clusterIPs)

thockin replied:
I think it's safe to assume that a user who changed ports meant to do so.  E.g.

CREATE { type=LB, ports[0] = (TCP, 8080) }
PATCH { type=ClusterIP, ports[0]=(TCP, 8000) }

I think that should NOT cause an error - we need to clear the NodePort.

khenidak commented on pkg/registry/core/service/strategy.go:
```diff
@@ -205,6 +207,127 @@ func topologyKeysInUse(svc *api.Service) bool {
+			}
+		}
+	}
+	// Every new port must not have a NodePort.
```
That does not make it same right? it confuses the user intent with (change type and add new ports)

thockin replied:
I'm saying that it is valid to change both things AS LONG AS the user did not specify a NodePort value.  This is closer to the create path, where we will barf if you set NodePort.  If the new state does not need a NodePort (e.g. ExternalName) and the new port does not have a NodePort, the user intent is clear.

thockin commented:
/retest

khenidak commented:
> this seems like a good addition for the narrowly scoped cases where we auto-allocate these fields
> 
> can move this question to Khal's PR if you want, but I'm curious if/when this is expanded to defaulted dual-stack fields (ipFamilies, ipFamilyPolicy, etc), under what circumstances would you envision clearing/re-populating those? something like "if the old and new field for {ipFamily,ipFamilyPolicy,clusterIPs} match values, and that value is not valid for the new service type"?

we have a similar thing for change to `ExternalName` type services. We don't compare old to new though. Also in strategy. We only clear. we leave repopulation to user and/or defaulting.

thockin commented:
/retest

thockin commented:
pushed with a net-simpler sameNodePorts impl.

Putting a hold until Kal and I decide sequencing this with his PR

/hold

liggitt commented:
/lgtm

> pushed with a net-simpler sameNodePorts impl.

doc updates and simpler sameNodePorts impl lgtm

> Putting a hold until Kal and I decide sequencing this with his PR

ack, will let you two wrangle order.

thockin committed:
Make some methods into non-methods

thockin committed:
Wipe some fields on service "type" updates

Service has had a problem since forever:

- User creates a service type=LoadBalancer
- We silently allocate them a NodePort
- User changes type to ClusterIP
- We fail the operation because they did not clear NodePort

They never asked for or used the NodePort!

Dual-stack introduced some dependent fields that get auto-wiped on
updates.  This carries it further.

If you squint, you can see Service as a big, messy discriminated union,
with type as the discriminator. Ignoring fields for non-selected
union-modes seems right.

This introduces the potential for an apply loop. Specifically, we will
accept YAML that we did not previously accept. Apply could see the
field in local YAML and not in the server and repeatedly try to patch it
in. But since that YAML is currently an error, it seems like a very low
risk. Almost nobody actually specifies their own NodePort values.

To mitigate this somewhat, we only auto-wipe on updates. The same YAML
would fail to create. This is a little inconsistent. We could
auto-wipe on create, too, at the risk of more potential impact.

To do this properly, we need to know the old and new values, which means
we can not do it in defaulting or conversion. So we do it in strategy.

This change also adds unit tests and updates e2e tests to rely on and
verify this behavior.

thockin commented:
Rebased this one HEAD which includes dual-stack support.

khenidak commented:
/lgtm
/approve

thockin commented:
/retest

thockin commented:
/retest
